pkg/pool-weighted/contracts/smart/LiquidityBootstrappingPool.sol

```diff
@@ -0,0 +1,501 @@
+
+        InputHelpers.ensureInputLengthMatch(numTokens, normalizedWeights.length);
+
+        // Ensure each normalized weight is above the minimum and find the token index of the maximum weight
+        uint256 normalizedSum = 0;
+        uint256 maxWeightTokenIndex = 0;
+        uint256 maxNormalizedWeight = 0;
+        for (uint8 i = 0; i < numTokens; i++) {
+            uint256 normalizedWeight = normalizedWeights[i];
+            _require(normalizedWeight >= _MIN_WEIGHT, Errors.MIN_WEIGHT);
+
+            _setNormalizedWeight(normalizedWeight, i);
+
+            normalizedSum = normalizedSum.add(normalizedWeight);
+            if (normalizedWeight > maxNormalizedWeight) {
+                maxWeightTokenIndex = i;
+                maxNormalizedWeight = normalizedWeight;
+            }
+        }
+        // Ensure that the normalized weights sum to ONE
+        _require(normalizedSum == FixedPoint.ONE, Errors.NORMALIZED_WEIGHT_INVARIANT);
+
+        // Initial value; can change later if weights change
+        _maxWeightTokenIndex = maxWeightTokenIndex;
+
+        publicSwapEnabled = initialPublicSwap;
+        minWeightChangeDuration = minDuration;
+    }
+
+    // External functions
+
+    /**
+     * @dev Can pause/unpause trading
+     */
+    function setPublicSwap(bool publicSwap) external onlyOwner whenNotPaused nonReentrant {
+        publicSwapEnabled = publicSwap;
+
+        emit PublicSwapSet(publicSwap);
+    }
+
+    /**
+     * @dev Schedule a gradual weight change, from the current normalizedWeights in storage, to
+     * the given endWeights, over startTime to endTime
+     */
+    function updateWeightsGradually(
+        uint256 startTime,
+        uint256 endTime,
+        uint256[] memory endWeights
+    ) external onlyOwner whenNotPaused nonReentrant {
+        // solhint-disable-next-line not-rely-on-time
+        uint256 currentTime = block.timestamp;
+
+        _require(currentTime < endTime, Errors.GRADUAL_UPDATE_TIME_TRAVEL);
+
+        // If called while a current weight change is ongoing, set starting point to current weights
+        _fixCurrentNormalizedWeights(currentTime);
+
+        if (currentTime > startTime) {
+            // This means the weight update should start ASAP
+            // Moving the start time up prevents a big jump/discontinuity in the weights
+            _gradualUpdateParams.startTime = currentTime;
+        } else {
+            _gradualUpdateParams.startTime = startTime;
+        }
+
+        // Enforce a minimum time over which to make the changes
+        // The also prevents endBlock <= startBlock
+        _require(
+            endTime.sub(_gradualUpdateParams.startTime) >= minWeightChangeDuration,
+            Errors.WEIGHT_CHANGE_TIME_BELOW_MIN
+        );
+
+        // Must specify normalized weights for all tokens
+        uint256 numTokens = _getTotalTokens();
+        InputHelpers.ensureInputLengthMatch(numTokens, endWeights.length);
```

Comment:
I'd move this check to the top of this function

Reply:
I can; was trying to save gas by doing the checks most likely to fail first, and the "error" ones last

Reply:
But yes; I can move it up before some of the heavier weight operations